base_layer/core/src/consensus/consensus_manager.rs

```diff
@@ -135,22 +112,28 @@ impl ConsensusManager {
         height: u64,
     ) -> Result<Difficulty, ConsensusManagerError>
     {
-        match self.access_diff_adj()?.as_ref() {
-            Some(v) => v
-                .get_target_difficulty_at_height(db, pow_algo, height)
-                .map_err(ConsensusManagerError::DifficultyAdjustmentManagerError),
-            None => Err(ConsensusManagerError::MissingDifficultyAdjustmentManager),
-        }
+        // TODO: store and use the target difficulty at horizon height as the initial difficulty for the
+        // LinearWeightedMovingAverage, then only the header set from horizon_height+1 to height need to be
+        // requested and processed.
```

Comment:
As mentioned here, a mechanism is required to store the previously calculated target difficulty for a historic height, the historic height(could be the horizon height) should be far enough into the history that it can't be reorged past that point. If an historic target difficulty is stored only the headers from that point in history need to be processed to calculate the updated target difficulty. Should this historic difficulty be stored using the blockchain db or should it be recalculated every time the node restarts? Using the blockchain db for storage will ensure that these functions remain stateless.

Reply:
I think we need to move to a separate store of each blocks target difficulty. Or at least a store quite a few blocks. If you look in the log file, how much the lwma is restarted for older blocks, and then its restarted from block 0 each time. This is super inefficient as this means we have to calculate the difficulty for currently about 9000 blocks if we only move 1 block back.

Reply:
I agree, that is exactly how it currently works and why this update is not much worse. I think storing the target difficulty at the horizon height and only calculating a new target difficulty from that point to the requested height will be more efficient and won't require any kind of reorg checking.